### ISSUE TITLE
Improve Synthesis Time By Multiprocessing

### ISSUE THREAD

**Issue by chazo1994** [CSTR-Edinburgh/merlin#347]
In my speech synthesis system built from Merlin toolkit, it take long time to generate speech from text. Most of the time used by World Vocoder and DNN generation module. So, to improve time delay, I propose an idea as follow:  in DNN decode (using theano), we generate acoustic features frame by frame, and push that features to vocoder to generate speech signal of frame. We do it in parallel (using multicore). Finally, we concatenate them to be a speech signal of entire input text. 
All of my idea is intergrate World vocoder and Acoustic DNN generation module to process the frame separately in multicore of multiprocessing. but How to generate feature of each frame with theano separatly with theano?

is My idea possible? How can I do it with theano and World vocoder? how to generate feature of each frame with theano separatly with theano? . Does anyone suggest me some solution to deploy this idea or  improve synthesis time?.

**Comment by trunglebka** [comment]
i have same problem :'(

**Comment by m-toman** [comment]
Are you using feedforward or recurrent networks?

A few things to consider:
Theano, when running on the CPU, typically parallelizes the matrix multiplications (http://deeplearning.net/software/theano/tutorial/multi_cores.html).

You can run MLPG as well as world in chunks and stream audio back, but have to be a bit careful.

Writing and reading intermediate files is costly. 

The MGC to spectrum function of SPTK is not optimized for this scenario.
You can find a couple optimizations for SPTK in my blog posts here:
http://www.neuratec.com/blog/?p=130
http://www.neuratec.com/blog/?p=99

**Comment by chazo1994** [comment]
Of course, I use feedforward neural network. I will consider your opinion carefully. thank you so much. I'm not sure, the conversion of MGC of each chunk to spectrum is right? it will generate right spectrum of speech?  And follow your opinion, My idea can be implemented, right? (online streaming)

**Comment by felipeespic** [comment]
The WORLD vocoder and SPTK are not optimised for streaming. You may end up allocating and deallocating a lot of memory on each frame processing.

**Comment by m-toman** [comment]
WORLD has a streaming method implemented, but it's not really documented. You can find an example here: https://github.com/mmorise/World/blob/master/test/test.cpp#L428-L438
But it's true this won't help you very much when you just call it from the command line.

Similarly with SPTK, it's quite a pain to fix all the memory leaks because it assumes to be run from the command line and then quit again (for example, lot of memory for windows is allocated and the only pointers moved to the middle of the block and never freed).

So what I do is:
Generate durations for 1-5 phones.
Generate acoustic features according to the durations.
For this chunk, transform features (mcg2sp etc.), MLPG and then "realtime" WORLD.
Stream back the waveform
Repeat.

That's a full C++ implementation, but perhaps you can find some middle-ground.

**Comment by chazo1994** [comment]
@m-toman Actually, I already chunked the label file ( after finish the generation of the HTS labels) and generate speech for each chunk without your tricky( with MLPG). But after concatenate waveform, I receive bad quality of voice with noise in the junctures between speech chunks (concatenation is not smooth).  I'm not sure your way can be better, because all of the function like Post filter, Vocoder,... are non linear function so the concatenation of chunks of sound is still a big problem.

**Comment by m-toman** [comment]
Yeah I don't run chunks in parallel, that way the quality is basically the same. So the last features of the previous chunk are available  when I start processing  the next one.
Just wanted to state that I do a bit of parallel processing inside each chunk (therefore better when the chunks are a bit larger) and that generally this is fast enough on most mobile devices for live synthesis.

**Comment by felipeespic** [comment]
> WORLD has a streaming method implemented, but it's not really documented. You can find an example here: https://github.com/mmorise/World/blob/master/test/test.cpp#L428-L438
> But it's true this won't help you very much when you just call it from the command line.

@m-toman Cool, I didn't know that WORLD had streaming mode implemented.

**Comment by npuichigo** [comment]
@chazo1994 By the way, can you tell me which file has memory leak problem in SPTK?
